src/connectors/impls/gpubsub/producer.rs

```diff
@@ -0,0 +1,234 @@
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct GpubSink {
+    config: Config,
+    url: Url<HttpsDefaults>,
+
+    client: Option<PublisherClient<InterceptedService<Channel, AuthInterceptor>>>,
+}
+
+#[async_trait::async_trait()]
+impl Sink for GpubSink {
+    async fn connect(&mut self, _ctx: &SinkContext, _attempt: &Attempt) -> Result<bool> {
+        let mut channel = Channel::from_shared(self.config.endpoint.clone())?
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout));
+        if self.url.scheme() == "https" {
+            let tls_config = ClientTlsConfig::new()
+                .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+                .domain_name(
+                    self.url
+                        .host_str()
+                        .ok_or_else(|| Status::unavailable("The endpoint is missing a hostname"))?
+                        .to_string(),
+                );
+
+            channel = channel.tls_config(tls_config)?;
+        }
+
+        let channel = channel.connect().await?;
+        #[allow(unused_assignments, unused_mut)]
+        let mut skip_authentication = false;
+        #[cfg(test)]
+        {
+            skip_authentication = self.config.skip_authentication;
+        }
+
+        let client = if skip_authentication {
+            info!("Skipping auth...");
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(|| Ok(Arc::new(String::new()))),
+                },
+            )
+        } else {
+            let token = Token::new()?;
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(move || {
+                        token.header_value().map_err(|_| {
+                            Status::unavailable("Failed to retrieve authentication token.")
+                        })
+                    }),
+                },
+            )
+        };
+
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self.client.as_mut().ok_or(ErrorKind::ClientNotAvailable(
+            "PubSub",
+            "The publisher is not connected",
+        ))?;
+
+        let mut messages = vec![];
+
+        for e in event.value_iter() {
+            messages.push(PubsubMessage {
+                data: serializer.serialize(e, event.ingest_ns)?[0].clone(),
+                attributes: HashMap::new(),
+                message_id: "".to_string(),
```

Comment:
this and the ordering_key should come from the event metadata, if available.

Reply:
The `message_id` is generated on the server side (I'll add a comment in the code). I'll get the `ordering_key` from the metadata though